crates/typst/src/image.rs

```diff
@@ -435,25 +497,64 @@ impl SvgFontLoader for WorldLoader<'_> {
             .clone()
     }
 
-    fn fallback_family(&self) -> Option<&str> {
-        self.fallback_family.as_deref()
+    fn fallback_families(&self) -> EcoVec<String> {
+        self.fallback_families.borrow().clone()
+    }
+
+    fn find_fallback(
+        &self,
+        text: &str,
+        like: Option<FontInfo>,
+        font: &usvg::Font,
+    ) -> Option<EcoString> {
+        let variant = FontVariant {
+            style: font.style.into(),
+            weight: FontWeight::from_number(font.weight),
+            stretch: font.stretch.into(),
+        };
+
+        let fallback: Option<EcoString> = self
+            .world
+            .book()
+            .select_fallback(like.as_ref(), variant, text)
+            .and_then(|id| self.world.font(id))
+            .map(|font| font.info().family.to_lowercase().as_str().into());
+
+        if let Some(fallback) = &fallback {
+            // TODO: Prevent duplicates
+            self.fallback_families.borrow_mut().push(fallback.to_string());
+        }
+
+        fallback
     }
 }
 
 /// Loads fonts for an SVG from a prepared list.
 #[derive(Default, Hash)]
 struct PreparedLoader {
     families: BTreeMap<EcoString, EcoVec<Font>>,
-    fallback_family: Option<EcoString>,
+    fallback_families: EcoVec<String>,
 }
 
 impl SvgFontLoader for PreparedLoader {
     fn load(&self, family: &str) -> EcoVec<Font> {
         self.families.get(family).cloned().unwrap_or_default()
     }
 
-    fn fallback_family(&self) -> Option<&str> {
-        self.fallback_family.as_deref()
+    fn fallback_families(&self) -> EcoVec<String> {
+        self.fallback_families.clone()
+    }
+
+    fn find_fallback(
+        &self,
+        _: &str,
+        _: Option<FontInfo>,
+        _: &usvg::Font,
+    ) -> Option<EcoString> {
+        // All fallbacks font should be included in 'families' already, so we will
+        // let usvg find it. As "Linux Libertine" must have failed to cover all glyphs,
+        // we can be sure that it will not actually be used as a fallback.
+        Some("Linux Libertine".into())
```

Comment:
I'm not a big fan of hard-coding `Linux Libertine` here. Is there a good reason not to just have `None` here?

I also noticed that this will sort-of defeat the point of the `SvgFontLoader` trait (memoizing across world and prepared loader) if fallback is involved since the `fallback_families` will differ across the `WorldLoader` and the `PreparedLoader`.

To ensure that things are memoized, both the `WorldLoader` and the `SvgFontLoader` would have to return the exact same things for each call of a tracked method. I guess the current version isn't too bad though since it will only happen when fallback is involved. Moreover, each of the two will still be memoized independently.